src/uct/ib/mlx5/gdaki/gdaki.c

```diff
@@ -653,6 +654,126 @@ uct_gdaki_md_check_uar(uct_ib_mlx5_md_t *md, CUdevice cuda_dev)
     return status;
 }
 
+typedef struct {
+    unsigned               index;
+    struct ibv_device      *dev;
+    ucs_sys_device_t       sys_dev;
+    ucs_sys_dev_distance_t dist;
+    int                    usecount;
+} uct_gdaki_dev_score_t;
+
+typedef struct {
+    ucs_sys_device_t sys_dev;
+    size_t           num;
+    CUdevice         gpus[8];
+} uct_gdaki_dev_matrix_elem_t;
+
+uct_gdaki_dev_matrix_elem_t *uct_gdaki_dev_matrix;
+size_t uct_gdaki_dev_matrix_length;
+int uct_gdaki_dev_matrix_initialized = 0;
+pthread_mutex_t uct_gdaki_dev_matrix_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static int uct_gdaki_dev_matrix_score(const void *pa, const void *pb, void *arg)
+{
+    const uct_gdaki_dev_score_t *a = pa;
+    const uct_gdaki_dev_score_t *b = pb;
+    int res = ucs_signum(a->dist.latency - b->dist.latency);
+    return res ? res : ucs_signum(a->usecount - b->usecount);
+}
+
+static ucs_status_t uct_gdaki_dev_matrix_init(void)
+{
+    struct ibv_device **device_list;
+    CUdevice device;
+    ucs_status_t status = UCS_OK;
+    int i, j, num_ibs, num_gpus;
+    uct_gdaki_dev_score_t *scores;
+    char *path_buffer;
+    ucs_sys_device_t gpu_dev;
+    const char *sysfs_path;
+    unsigned ib_per_gpu;
+    uct_gdaki_dev_matrix_elem_t *ibdesc;
+
+    status = ucs_string_alloc_path_buffer(&path_buffer, "path_buffer");
+    if (status != UCS_OK) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    device_list = ibv_get_device_list(&num_ibs);
+    if (device_list == NULL) {
+        status = UCS_ERR_IO_ERROR;
+        goto out_buff;
+    }
+
+    uct_gdaki_dev_matrix = ucs_calloc(num_ibs, sizeof(*uct_gdaki_dev_matrix),
+                                      "dev matrix");
+    if (uct_gdaki_dev_matrix == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out_dev;
+    }
+
+    scores = ucs_calloc(num_ibs, sizeof(*scores), "dev scores");
+    if (scores == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out_dev;
+    }
+
+    for (i = 0; i < num_ibs; i++) {
+        ibdesc        = &uct_gdaki_dev_matrix[i];
+        scores[i].dev = device_list[i];
+
+        sysfs_path = ucs_topo_resolve_sysfs_path(scores[i].dev->ibdev_path,
+                                                 path_buffer);
+        scores[i].sys_dev = ucs_topo_get_sysfs_dev(ibv_get_device_name(
```

Comment:
Not specific for this PR:
Code for getting sys_dev from path & name repeats in more places in UCX.
Maybe we can add common function to ucs_topo (can be in future PR).